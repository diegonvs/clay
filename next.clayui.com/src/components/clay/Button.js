/**
 * Â© 2019 Liferay, Inc. <https://liferay.com>
 *
 * SPDX-License-Identifier: BSD-3-Clause
 */

import ClayButton from '@clayui/button';
import ClayIcon from '@clayui/icon';
import Editor from './Editor';
import React from 'react';

const spritemap = '/images/icons/icons.svg';

export const ButtonBlock = () => {
	const scope = {ClayButton, spritemap};
	const code = `const Component = () => {

	return (
        <ClayButton displayType="primary" block>
			Button Block
		</ClayButton>
	);
}

render(<Component />)`;

	return <Editor code={code} scope={scope} />;
};

export const ButtonDisplayTypes = () => {
	const scope = {ClayButton, spritemap};
	const code = `const Component = () => {

	return (
		<>
			<ClayButton displayType="primary">
				Button Primary
			</ClayButton>
			<ClayButton displayType="secondary">
				Button Secondary
			</ClayButton>
			<ClayButton displayType="link">
				Button Link
			</ClayButton>
			<ClayButton displayType="unstyled">
				Button Unstyled
			</ClayButton>
		</>
	);
}

render(<Component />)`;

	return <Editor code={code} scope={scope} />;
};

export const ButtonGroup = () => {
	const scope = {ClayButton, spritemap};
	const code = `const Component = () => {

	return (
        <ClayButton.Group>
			<ClayButton>{'This'}</ClayButton>
			<ClayButton displayType="secondary">{'is'}</ClayButton>
			<ClayButton>{'a'}</ClayButton>
			<ClayButton displayType="secondary">{'button'}</ClayButton>
			<ClayButton>{'group.'}</ClayButton>
		</ClayButton.Group>
	);
}

render(<Component />)`;

	return <Editor code={code} scope={scope} />;
};

export const ButtonMonospaced = () => {
	const scope = {ClayButton, ClayIcon, spritemap};
	const code = `const Component = () => {

	return (
		<ClayButton displayType="primary" monospaced>
			<ClayIcon spritemap={spritemap} symbol="plus"></ClayIcon>
		</ClayButton>
	);
}

render(<Component />)`;

	return <Editor code={code} scope={scope} />;
};

export const ButtonWithStar = () => {
	const scope = {ClayButton, ClayIcon, spritemap};
	const code = `const Component = () => {

	return (
		<ClayButton displayType="primary">
			<ClayIcon spritemap={spritemap} symbol="bookmarks"></ClayIcon> Bookmark
		</ClayButton>
	);
}

render(<Component />)`;

	return <Editor code={code} scope={scope} />;
};

export const ButtonSmall = () => {
	const scope = {ClayButton, ClayIcon, spritemap};
	const code = `const Component = () => {

	return (
		<ClayButton displayType="primary" small>
			<ClayIcon spritemap={spritemap} symbol="plus"></ClayIcon>
		</ClayButton>
	);
}

render(<Component />)`;

	return <Editor code={code} scope={scope} />;
};
