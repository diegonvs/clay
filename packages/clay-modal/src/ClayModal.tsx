/**
 * Â© 2019 Liferay, Inc. <https://liferay.com>
 *
 * SPDX-License-Identifier: BSD-3-Clause
 */

import Button from '@clayui/button';
import classNames from 'classnames';
import Icon from '@clayui/icon';
import React, {FunctionComponent, useEffect, useRef, useState} from 'react';
import {createPortal} from 'react-dom';
import {Footer, FooterContainer, FooterContainerProps} from './Footer';
import {useModal} from './Hook';

type Size = 'full-screen' | 'lg' | 'sm';

type Status = 'danger' | 'info' | 'success' | 'warning';

interface Props extends React.HTMLAttributes<HTMLDivElement> {
	/**
	 * Callback called to close the modal.
	 */
	onClose: () => void;

	/**
	 * Renders the footer with the FooterContainer component utilities.
	 */
	renderFooter?: React.ReactElement;

	/**
	 * The size of element modal.
	 */
	size?: Size;

	/**
	 * The path to the SVG spritemap file containing the icons.
	 */
	spritemap?: string;

	/**
	 * Status messages.
	 */
	status?: Status;

	/**
	 * Title of the modal.
	 */
	title?: string;

	/**
	 * Url to place an iframe in the body of the modal.
	 */
	url?: string;

	/**
	 * Flag to indicate if the modal is open or closed.
	 */
	visible: boolean;
}

const ICON_MAP = {
	danger: 'exclamation-full',
	info: 'info-circle',
	success: 'check-circle',
	warning: 'question-circle-full',
};

const Portal = ({children}: {children: React.ReactNode}) => {
	const portalRef = useRef(document.createElement('div'));
	const elToMountTo = document && document.body;

	useEffect(() => {
		elToMountTo.appendChild(portalRef.current);
		return () => {
			elToMountTo.removeChild(portalRef.current);
		};
	}, [elToMountTo]);

	return createPortal(children, portalRef.current);
};

const delay = (fn: Function) => {
	const callbackId = setTimeout(() => {
		fn();
		clearTimeout(callbackId);
	}, 100);
};

const ClayModal: FunctionComponent<Props> & {
	FooterContainer: FunctionComponent<FooterContainerProps>;
} = ({
	children,
	className,
	onClose = () => {},
	renderFooter,
	size,
	spritemap,
	status,
	title,
	url,
	visible = false,
	...otherProps
}) => {
	/**
	 * It is set to keep track of modal visibility and ensure that classes
	 * are removed when necessary.
	 */
	const [visibleModal, setVisibleModal] = useState<boolean>(visible);

	const [visibleClassShow, setVisibleClassShow] = useState<boolean>(false);

	const modalDialogElementRef = useRef<HTMLDivElement | null>(null);

	/**
	 * A hook that takes care of controlling click, keyup and keydown events
	 * respectively close the modal after a click on the overlay, close the
	 * modal by pressing the ESC key and control the focus within the Modal.
	 */
	useModal(modalDialogElementRef, visible, onClose);

	/**
	 * Synchronize the visible of the props with the internal value to update
	 * the states with delay, this ensures that the animation occurs when
	 * opening and closing the modal.
	 */
	useEffect(() => {
		if (visible) {
			document.body.classList.add('modal-open');
			setVisibleModal(visible);
			delay(() => setVisibleClassShow(visible));
		} else {
			document.body.classList.remove('modal-open');
			setVisibleClassShow(visible);
			delay(() => setVisibleModal(visible));
		}
	}, [visible]);

	return (
		<>
			{visibleModal && (
				<Portal>
					<div
						className={classNames('modal-backdrop fade', {
							show: visibleClassShow,
						})}
					/>
				</Portal>
			)}
			<div
				{...otherProps}
				className={classNames('fade modal', className, {
					'd-block': visibleModal,
					show: visibleClassShow,
				})}
			>
				<div
					className={classNames('modal-dialog', {
						[`modal-${size}`]: size,
						[`modal-${status}`]: status,
					})}
					ref={modalDialogElementRef}
					tabIndex={-1}
				>
					<div className="modal-content">
						{title && (
							<div className="modal-header">
								{status && (
									<div className="modal-title-indicator">
										<Icon
											spritemap={spritemap}
											symbol={ICON_MAP[status]}
										/>
									</div>
								)}
								<div className="modal-title">{title}</div>
								<Button
									aria-label="close"
									className="close"
									displayType="unstyled"
									onClick={onClose}
								>
									<Icon
										spritemap={spritemap}
										symbol="times"
									/>
								</Button>
							</div>
						)}
						<div
							className={classNames('modal-body', {
								'modal-body-iframe': url,
							})}
						>
							{url ? <iframe src={url} title={url} /> : children}
						</div>
						{renderFooter && <Footer>{renderFooter}</Footer>}
					</div>
				</div>
			</div>
		</>
	);
};

ClayModal.FooterContainer = FooterContainer;

export default ClayModal;
