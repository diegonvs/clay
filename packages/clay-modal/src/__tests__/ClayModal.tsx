/**
 * Â© 2019 Liferay, Inc. <https://liferay.com>
 *
 * SPDX-License-Identifier: BSD-3-Clause
 */

import * as React from 'react';
import * as TestRenderer from 'react-test-renderer';
import Button from '@clayui/button';
import ClayModal from '../ClayModal';
import ReactDOM from 'react-dom';

const spritemap = 'icons.svg';

describe('ClayModal', () => {
	beforeAll(() => {
		// @ts-ignore
		jest.spyOn(global.console, 'error').mockImplementation(() => {});
		// @ts-ignore
		ReactDOM.createPortal = jest.fn((element, node) => {
			return element;
		});
	});

	afterEach(() => {
		// @ts-ignore
		ReactDOM.createPortal.mockClear();
		// @ts-ignore
		// eslint-disable-next-line no-console
		console.error.mockClear();
	});

	it('renders', () => {
		const testRenderer = TestRenderer.create(
			<ClayModal
				onClose={() => {}}
				spritemap={spritemap}
				visible={false}
			/>
		);

		expect(testRenderer.toJSON()).toMatchSnapshot();
	});

	it('should renders with visible `true`', () => {
		const testRenderer = TestRenderer.create(
			<ClayModal onClose={() => {}} spritemap={spritemap} visible />
		);

		expect(testRenderer.toJSON()).toMatchSnapshot();
	});

	it('should renders with title', () => {
		const testRenderer = TestRenderer.create(
			<ClayModal
				onClose={() => {}}
				spritemap={spritemap}
				title="Foo"
				visible={false}
			/>
		);

		expect(testRenderer.toJSON()).toMatchSnapshot();
	});

	it('should renders with size', () => {
		const testRenderer = TestRenderer.create(
			<ClayModal
				onClose={() => {}}
				size="lg"
				spritemap={spritemap}
				visible={false}
			/>
		);

		expect(testRenderer.toJSON()).toMatchSnapshot();
	});

	it('should renders with status', () => {
		const testRenderer = TestRenderer.create(
			<ClayModal
				onClose={() => {}}
				spritemap={spritemap}
				status="success"
				visible={false}
			/>
		);

		expect(testRenderer.toJSON()).toMatchSnapshot();
	});

	it('should renders with url', () => {
		const testRenderer = TestRenderer.create(
			<ClayModal
				onClose={() => {}}
				spritemap={spritemap}
				url="http://localhost"
				visible={false}
			/>
		);

		expect(testRenderer.toJSON()).toMatchSnapshot();
	});

	it('should renders with one element in renderFooter', () => {
		const testRenderer = TestRenderer.create(
			<ClayModal
				onClose={() => {}}
				renderFooter={
					<ClayModal.FooterContainer>
						<Button>{'Bar'}</Button>
					</ClayModal.FooterContainer>
				}
				spritemap={spritemap}
				url="http://localhost"
				visible={false}
			/>
		);

		expect(testRenderer.toJSON()).toMatchSnapshot();
	});

	it('should renders with two element in renderFooter', () => {
		const testRenderer = TestRenderer.create(
			<ClayModal
				onClose={() => {}}
				renderFooter={
					<>
						<ClayModal.FooterContainer>
							<Button>{'Bar'}</Button>
						</ClayModal.FooterContainer>
						<ClayModal.FooterContainer>
							<Button>{'Baz'}</Button>
						</ClayModal.FooterContainer>
					</>
				}
				spritemap={spritemap}
				url="http://localhost"
				visible={false}
			/>
		);

		expect(testRenderer.toJSON()).toMatchSnapshot();
	});

	it('should render but emit error when wrapping FooterContainer with HTML tag', () => {
		const testRenderer = TestRenderer.create(
			<ClayModal
				onClose={() => {}}
				renderFooter={
					<div>
						<ClayModal.FooterContainer>
							<Button>{'Bar'}</Button>
						</ClayModal.FooterContainer>
					</div>
				}
				spritemap={spritemap}
				url="http://localhost"
				visible={false}
			/>
		);

		// eslint-disable-next-line no-console
		expect(console.error).toBeCalled();
	});

	it('should render but emit error when using more than two <FooterContainer /> components', () => {
		const testRenderer = TestRenderer.create(
			<ClayModal
				onClose={() => {}}
				renderFooter={
					<>
						<ClayModal.FooterContainer>
							<Button>{'Foo'}</Button>
						</ClayModal.FooterContainer>
						<ClayModal.FooterContainer>
							<Button>{'Bar'}</Button>
						</ClayModal.FooterContainer>
						<ClayModal.FooterContainer>
							<Button>{'Baz'}</Button>
						</ClayModal.FooterContainer>
					</>
				}
				spritemap={spritemap}
				url="http://localhost"
				visible={false}
			/>
		);

		// eslint-disable-next-line no-console
		expect(console.error).toBeCalled();
	});
});
