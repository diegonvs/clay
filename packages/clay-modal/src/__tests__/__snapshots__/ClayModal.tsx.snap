// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ClayModal renders 1`] = `
<div
  className="fade modal"
>
  <div
    className="modal-dialog"
    tabIndex={-1}
  >
    <div
      className="modal-content"
    >
      <div
        className="modal-body"
      />
    </div>
  </div>
</div>
`;

exports[`ClayModal should renders with one element in renderFooter 1`] = `
<div
  className="fade modal"
>
  <div
    className="modal-dialog"
    tabIndex={-1}
  >
    <div
      className="modal-content"
    >
      <div
        className="modal-body modal-body-iframe"
      >
        <iframe
          src="http://localhost"
          title="http://localhost"
        />
      </div>
      <div
        className="modal-footer"
      >
        <div
          className="modal-item-first"
        >
          <div
            className="btn-group"
          >
            <div
              className="btn-group-item"
            >
              <button
                className="btn btn-primary"
                type="button"
              >
                Bar
              </button>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`ClayModal should renders with size 1`] = `
<div
  className="fade modal"
>
  <div
    className="modal-dialog modal-lg"
    tabIndex={-1}
  >
    <div
      className="modal-content"
    >
      <div
        className="modal-body"
      />
    </div>
  </div>
</div>
`;

exports[`ClayModal should renders with status 1`] = `
<div
  className="fade modal"
>
  <div
    className="modal-dialog modal-success"
    tabIndex={-1}
  >
    <div
      className="modal-content"
    >
      <div
        className="modal-body"
      />
    </div>
  </div>
</div>
`;

exports[`ClayModal should renders with title 1`] = `
<div
  className="fade modal"
>
  <div
    className="modal-dialog"
    tabIndex={-1}
  >
    <div
      className="modal-content"
    >
      <div
        className="modal-header"
      >
        <div
          className="modal-title"
        >
          Foo
        </div>
        <button
          aria-label="close"
          className="close btn btn-unstyled"
          onClick={[Function]}
          type="button"
        >
          <svg
            className="lexicon-icon lexicon-icon-times"
            role="presentation"
          >
            <use
              xlinkHref="icons.svg#times"
            />
          </svg>
        </button>
      </div>
      <div
        className="modal-body"
      />
    </div>
  </div>
</div>
`;

exports[`ClayModal should renders with two element in renderFooter 1`] = `
<div
  className="fade modal"
>
  <div
    className="modal-dialog"
    tabIndex={-1}
  >
    <div
      className="modal-content"
    >
      <div
        className="modal-body modal-body-iframe"
      >
        <iframe
          src="http://localhost"
          title="http://localhost"
        />
      </div>
      <div
        className="modal-footer"
      >
        <div
          className="modal-item-first"
        >
          <div
            className="btn-group"
          >
            <div
              className="btn-group-item"
            >
              <button
                className="btn btn-primary"
                type="button"
              >
                Bar
              </button>
            </div>
          </div>
        </div>
        <div
          className="modal-item-last"
        >
          <div
            className="btn-group"
          >
            <div
              className="btn-group-item"
            >
              <button
                className="btn btn-primary"
                type="button"
              >
                Baz
              </button>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`ClayModal should renders with url 1`] = `
<div
  className="fade modal"
>
  <div
    className="modal-dialog"
    tabIndex={-1}
  >
    <div
      className="modal-content"
    >
      <div
        className="modal-body modal-body-iframe"
      >
        <iframe
          src="http://localhost"
          title="http://localhost"
        />
      </div>
    </div>
  </div>
</div>
`;

exports[`ClayModal should renders with visible \`true\` 1`] = `
Array [
  <div
    className="modal-backdrop fade"
  />,
  <div
    className="fade modal d-block"
  >
    <div
      className="modal-dialog"
      tabIndex={-1}
    >
      <div
        className="modal-content"
      >
        <div
          className="modal-body"
        />
      </div>
    </div>
  </div>,
]
`;
